ltdconveyor/s3upload.py

```diff
@@ -0,0 +1,395 @@
+                continue
+            obj_dirname = os.path.dirname(obj.key)
+            if obj_dirname == prefix:
+                # object is at root of directory
+                filenames.append(os.path.relpath(obj.key,
+                                                 start=prefix))
+        return filenames
+
+    def list_dirnames_in_directory(self, dirname):
+        """List all names of directories that exist at the root of this
+        bucket directory.
+
+        Note that *directories* don't exist in S3; rather directories are
+        inferred from path names.
+
+        Parameters
+        ----------
+        dirname : `str`
+            Directory name in the bucket relative to ``bucket_root``.
+
+        Returns
+        -------
+        dirnames : `list`
+            List of directory names (`str`), relative to ``bucket_root/``,
+            that exist at the root of ``dirname``.
+        """
+        prefix = self._create_prefix(dirname)
+        dirnames = []
+        for obj in self._bucket.objects.filter(Prefix=prefix):
+            dirname = os.path.dirname(obj.key)
+            # if the object is a directory redirect, make it look like a dir
+            if dirname == '':
+                dirname = obj.key + '/'
+            rel_dirname = os.path.relpath(dirname, start=prefix)
+            dir_parts = rel_dirname.split('/')
+            if len(dir_parts) == 1:
+                dirnames.append(dir_parts[0])
+        dirnames = list(set(dirnames))
+        if '.' in dirnames:
+            dirnames.remove('.')
+        return dirnames
+
+    def _create_prefix(self, dirname):
+        if dirname in ('.', '/'):
+            dirname = ''
+        # Strips trailing slash from dir prefix for comparisons
+        # os.dirname() returns directory names without a trailing /
+        prefix = os.path.join(self._bucket_root, dirname)
+        if prefix.endswith('/'):
+            prefix = prefix.rstrip('/')
+        return prefix
+
+    def delete_file(self, filename):
+        """Delete a file from the bucket.
+
+        Parameters
+        ----------
+        filename : `str`
+            Name of the file, relative to ``bucket_root/``.
+        """
+        key = os.path.join(self._bucket_root, filename)
+        objects = list(self._bucket.objects.filter(Prefix=key))
+        for obj in objects:
+            obj.delete()
+
+    def delete_directory(self, dirname):
+        """Delete a directory (and contents) from the bucket.
+
+        Parameters
+        ----------
+        dirname : `str`
+            Name of the directory, relative to ``bucket_root/``.
+        """
+        key = os.path.join(self._bucket_root, dirname)
+        if not key.endswith('/'):
+            key += '/'
+
+        delete_keys = {'Objects': []}
+        key_objects = [{'Key': obj.key}
+                       for obj in self._bucket.objects.filter(Prefix=key)]
+        assert len(key_objects) > 0
+        delete_keys['Objects'] = key_objects
```

Comment:
Suggest combining the assignment with the initialization above OR moving the initialization of `delete_keys` down to this lin.